src/Features/CSharp/Portable/CodeRefactorings/ConvertIfToSwitch/CSharpConvertToSwitchCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,293 @@
+                            {
+                                return null;
+                            }
+
+                            switch (node.Pattern)
+                            {
+                                case ConstantPatternSyntax p:
+                                    return new Pattern.ByValue(p.Expression);
+
+                                case DeclarationPatternSyntax p
+                                    when p.Designation is SingleVariableDesignationSyntax designation:
+                                    return new Pattern.ByType(p.Type, designation.Identifier);
+
+                                default:
+                                    return null;
+                            }
+                        }
+
+                    case SyntaxKind.LogicalAndExpression:
+                        {
+                            ExpressionSyntax GetLeftmostCondition(ExpressionSyntax syntaxNode)
+                            {
+                                syntaxNode = syntaxNode.WalkDownParentheses();
+                                while (syntaxNode.IsKind(SyntaxKind.LogicalAndExpression))
+                                {
+                                    syntaxNode = ((BinaryExpressionSyntax)syntaxNode).Left.WalkDownParentheses();
+                                }
+
+                                return syntaxNode;
+                            }
+
+                            var node = (BinaryExpressionSyntax)operand;
+                            var leftmost = GetLeftmostCondition(node);
+                            switch (leftmost.Kind())
+                            {
+                                case SyntaxKind.IsPatternExpression:
+                                case SyntaxKind.IsExpression:
+                                case SyntaxKind.EqualsExpression:
+                                    break;
+
+                                default:
+                                    return null;
+                            }
+
+                            var p = GetPattern(leftmost);
+                            if (p == null)
+                            {
+                                return null;
+                            }
+
+                            var condition = (leftmost.Parent.Parent as BinaryExpressionSyntax)
+                                ?.WithLeft(((BinaryExpressionSyntax)leftmost.Parent).Right);
+
+                            return new Pattern.Guarded(p, (condition ?? node.Right).WalkDownParentheses());
+                        }
+
+                    default:
+                        return null;
+                }
+            }
+
+            StatementSyntax elseBody;
+            var sectionData = new List<(List<Pattern> patterns, StatementSyntax body)>();
+            var currentIfStatement = ifStatement;
+            while (true)
+            {
+                var operands = GetLogicalOrExpressionOperands(currentIfStatement.Condition).ToImmutableArray();
```

Comment:
is converting to an immutable array actaully necessary.  seems like you just iterate over this thing once.